core/vm/lfvm/interpreter.go

```diff
@@ -355,9 +378,15 @@ func runWithStatistics(c *context) {
 }
 
 func step(c *context) {
-	// TODO: remove this implicit stop at the end of a code section
-	if int(c.pc) >= len(c.code) {
-		opStop(c)
+
+	// If the interpreter is operating in readonly mode, make sure no
+	// state-modifying operation is performed. The 3rd stack item
+	// for a call operation is the value. Transferring value from one
+	// account to the others means the state is modified and should also
+	// return with an error.
+	if c.readOnly && (isWriteInstruction(c.code[c.pc].opcode) || (c.code[c.pc].opcode == CALL && c.stack.Back(2).Sign() != 0)) {
```

Comment:
It might be beneficial to move this `readOnly` check into write instruction implementations to avoid testing this for each instruction - in particular for the most frequent ones, like Push and Swap instructions.